Marlin/src/gcode/feature/trinamic/M919.cpp

```diff
@@ -0,0 +1,257 @@
+/**
+ * Marlin 3D Printer Firmware
+ * Copyright (c) 2021 MarlinFirmware [https://github.com/MarlinFirmware/Marlin]
+ *
+ * Based on Sprinter and grbl.
+ * Copyright (c) 2011 Camiel Gubbels / Erik van der Zalm
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "../../../inc/MarlinConfig.h"
+
+#if HAS_TRINAMIC_CONFIG
+
+#include "../../gcode.h"
+#include "../../../feature/tmc_util.h"
+#include "../../../module/stepper/indirection.h"
+
+#define DEBUG_OUT ENABLED(MARLIN_DEV_MODE)
+#include "../../../core/debug_out.h"
+
+/**
+ * M919: Set TMC stepper drive chopper times
+ *
+ * Parameters:
+ *   XYZIJKE...
+ *   I[index]    - Axis sub-index (Omit or 0 for X, Y, Z, E0; 1 for X2, Y2, Z2, E1; 2 for Z3, E2; 3 for Z4, E3; 4 for E4; 5 for E5; 6 for E6; 7 for E7)
+ *   T[index]    - Extruder index (Zero-based. Omit for E0 only.)
+ *   O           - time-off         [ 1..15]
+ *   V           - hysteresis_end   [-3..12]
+ *   S           - hysteresis_start [ 1...8]
+ * 
+ * With no parameters report chopper times for all axis
+ */
+
+void GcodeSuite::M919() {
+  #define TMC_SAY_CHOPPER_TIME(Q) tmc_print_chopper_time(stepper##Q)
+  #define TMC_SET_CHOPPER_TIME(Q) stepper##Q.set_chopper_times(time_off, hysteresis_end, hysteresis_start)
+
+  bool report = true;
+  uint8_t time_off;
+  uint8_t hysteresis_start;
+  int8_t hysteresis_end;
+
+  #if AXIS_IS_TMC(X) || AXIS_IS_TMC(X2) || AXIS_IS_TMC(Y) || AXIS_IS_TMC(Y2) || AXIS_IS_TMC(Z) || AXIS_IS_TMC(Z2) || AXIS_IS_TMC(Z3) || AXIS_IS_TMC(Z4)
+    const uint8_t index = parser.byteval('I');
+  #endif
+    
+  static constexpr chopper_timing_t chopper_timing = CHOPPER_TIMING;
+  time_off = chopper_timing.toff;
+  hysteresis_end = chopper_timing.hend;
+  hysteresis_start = chopper_timing.hstrt;
```

Comment:
I'm not sure what you mean by that

Reply:
I see. So instead of copying values to `time_off`, `hysteresis_end`, and `hysteresis_start` just directly use the `chopper_timing.toff`, `.hend`, and `.hstrt`. I'll take a quick look and confirm that it's an option.